cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs

```diff
@@ -0,0 +1,656 @@
+            destination = outputDir </> "genesis-shelley.json"
+        liftIO $ LBS.writeFile destination $ encode newTemplate
+        pure newTemplate
+
+  -- Genesis keys
+  -- {0 -> genesis-keys/genesis0.vkey, 1 -> genesis-keys/genesis1.vkey, ...}
+  let mkPaths :: String -> String -> String -> Map Int FilePath =
+        \parentDir dirPrefix filename ->
+          fromList
+          [(fromIntegral idx,
+            outputDir </> parentDir </> (dirPrefix <> show idx) </> filename)
+           | idx <- [1 .. numGenesisKeys]]
+      -- {0 -> genesis-keys/genesis0/key.vkey, 1 -> genesis-keys/genesis1/key.vkey, ...}
+      genesisVKeysPaths = mkPaths "genesis-keys" "genesis" "key.vkey"
+      -- {0 -> delegate-keys/delegate0/key.vkey, 1 -> delegate-keys/delegate1/key.vkey, ...}
+      delegateKeys = mkPaths "delegate-keys" "delegate" "key.vkey"
+      -- {0 -> delegate-keys/delegate0/vrf.vkey, 1 -> delegate-keys/delegate1/vrf.vkey, ...}
+      delegateVrfKeys = mkPaths "delegate-keys" "delegate" "vrf.vkey"
+
+  forM_ [ 1 .. numGenesisKeys ] $ \index -> do
+    createGenesisKeys (genesisDir </> ("genesis" <> show index))
+    createDelegateKeys keyOutputFormat (outputDir </> "delegate-keys" </> ("delegate" <> show index))
+
+  -- UTxO keys
+  let utxoKeys = [outputDir </> "utxo-keys" </> ("utxo" <> show index) </> "utxo.vkey"
+                  | index <-  [ 1 .. numUtxoKeys ]]
+  forM_ [ 1 .. numUtxoKeys ] $ \index ->
+    createUtxoKeys $ outputDir </> "utxo-keys" </> ("utxo" <> show index)
+
+  let mayStakePoolRelays = Nothing -- TODO @smelc temporary?
+
+  -- Pools
+  poolParams <- forM [ 1 .. numPools ] $ \index -> do
+    let poolsDir = outputDir </> "pools-keys"
+        poolDir = poolsDir </> ("pool" <> show index)
+
+    createPoolCredentials keyOutputFormat poolDir Nothing
+    buildPoolParams network poolDir Nothing (fromMaybe mempty mayStakePoolRelays)
+
+  -- Stake delegators
+  let (delegsPerPool, delegsRemaining) = divMod numStakeDelegators numPools
+      delegsForPool poolIx = if delegsRemaining /= 0 && poolIx == numPools
+        then delegsPerPool
+        else delegsPerPool + delegsRemaining
+      distribution = [pool | (pool, poolIx) <- zip poolParams [1 ..], _ <- [1 .. delegsForPool poolIx]]
+
+  g <- Random.getStdGen
+
+  -- Distribute M delegates across N pools:
+  delegations <- liftIO $ Lazy.forStateM g distribution $ flip computeInsecureDelegation network
+
+  genDlgs <- readGenDelegsMap genesisVKeysPaths delegateKeys delegateVrfKeys
+  nonDelegAddrs <- readInitialFundAddresses utxoKeys network
+  start <- maybe (SystemStart <$> getCurrentTimePlus30) pure systemStart
+
+  stuffedUtxoAddrs <- liftIO $ Lazy.replicateM (fromIntegral numStuffedUtxo) genStuffedAddress
+
+  let stake = second Ledger.ppId . mkDelegationMapEntry <$> delegations
+      stakePools = [ (Ledger.ppId poolParams', poolParams') | poolParams' <- snd . mkDelegationMapEntry <$> delegations ]
```

Comment:
You're right, but this doesn't matter, because the list is turned into a map in `updateCreateStakedOutputTemplate`.

Nevertheless I made this clear by using a `Set` rather than a `List` (in an additional commit :+1:)